### PR TITLE
update and clarify addr_of docs

### LINKED ISSUE #114902: `addr_of[_mut]!` docs should be more precise about what's sound
_This example comes from [this URLO thread](https://users.rust-lang.org/t/how-to-use-addr-of-to-create-an-unaligned-pointer/98524), which in turn was inspired by [this zerocopy issue](https://github.com/google/zerocopy/issues/196)._

The [`addr_of!`](https://doc.rust-lang.org/stable/std/ptr/macro.addr_of.html) and [`addr_of_mut!`](https://doc.rust-lang.org/stable/std/ptr/macro.addr_of.html) docs imply that they can handle unaligned pointers so long as a reference is never materialized from that pointer (which would be insta-UB). However, running the following program under Miri fails:

```rust
fn main() {
    #[repr(C, packed)]
    struct Unalign<T>(T);
    
    #[repr(C)]
    struct Foo {
        a: u8,
        b: u16,
    }
    
    // Has alignment `align_of::<T>()`, and the `Unalign<T>`
    // is at byte offset 1; so long as `align_of::<T>() > 1`,
    // the contained `T` is misaligned.
    #[repr(C)]
    struct Misalign<T>(u8, Unalign<T>, [T; 0]);
    
    let u = Misalign(0, Unalign(Foo{ a: 1, b: 2 }), []);
    let u_ptr: *const Unalign<Foo> = &u.1;
    // Sound because `Unalign` contains a `T` and nothing else.
    let f_ptr: *const Foo = u_ptr.cast();
    // Should be sound because we never construct a reference.
    let addr_of_b: *const u16 = unsafe { core::ptr::addr_of!((*f_ptr).b) };
    println!("{:?}", addr_of_b);
}
```

Here's the failure:

```text
error: Undefined Behavior: accessing memory with alignment 1, but alignment 2 is required
  --> src/main.rs:22:42
   |
22 |     let addr_of_b: *const u16 = unsafe { core::ptr::addr_of!((*f_ptr).b) };
   |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ accessing memory with alignment 1, but alignment 2 is required
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: BACKTRACE:
   = note: inside `main` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:1980:5: 1980:22
   = note: this error originates in the macro `core::ptr::addr_of` (in Nightly builds, run with -Z macro-backtrace for more info)
```

@cuviper [pointed out my issue](https://users.rust-lang.org/t/how-to-use-addr-of-to-create-an-unaligned-pointer/98524/2?u=joshlf):

> The example shows an unaligned *field* (since it's packed), but that doesn't excuse everything. "Note, however, that the `expr` in `addr_of!(expr)` is still subject to all the usual rules." -- I think this would include alignment for `(*f_ptr)`, although the docs only call out null pointers.
>
> Miri is happy with `addr_of!((*u_ptr).0.b)`.

The docs strongly imply that this would be sound. It does this in two ways. First, when contrasting references and raw pointers, it specifically calls out that references need to be aligned:

> Creating a reference with `&`/`&mut` is only allowed if the pointer is properly aligned and points to initialized data. For cases where those requirements do not hold, raw pointers should be used instead.

Second, it seems to enumerate the cases in which `addr_of!` would be unsound, and operating on unaligned pointers isn't on the list:

> Note, however, that the `expr` in `addr_of!(expr)` is still subject to all the usual rules. In particular, `addr_of!(*ptr::null())` is Undefined Behavior because it dereferences a null pointer.

IMO the docs as currently written are both vague and misleading with respect to what's allowed. It would be good if the docs were more precise on exactly what is and isn't allowed. I was pretty confident in the soundness of the code I'd written, and only discovered the issue thanks to Miri.

### PR DESCRIPTION
**Pull request by RalfJung**
This updates the docs to match https://github.com/rust-lang/reference/pull/1387. Cc @rust-lang/opsem 

@chorman0773 not sure if you had anything else you wanted to say here, I'd be happy to get your feedback. :)

Fixes https://github.com/rust-lang/rust/issues/114902, so Cc @joshlf

### CODE REVIEW

**Comment by chorman0773** [comment]
Appologies for forgetting to work on this.

I'd probably also explicitly note that `expr` applies `Deref` and `DerefMut` (likewise `Index` and `IndexMut`) when applicable, as that's a fairly significant footgun.

**Comment by RalfJung** [comment]
Good point, done.

File: library/core/src/ptr/mod.rs [hunk 1]
```diff
@@ -1989,9 +1989,18 @@ impl<F: FnPtr> fmt::Debug for F {
 /// as all other references. This macro can create a raw pointer *without* creating
 /// a reference first.
 ///
-/// Note, however, that the `expr` in `addr_of!(expr)` is still subject to all
-/// the usual rules. In particular, `addr_of!(*ptr::null())` is Undefined
-/// Behavior because it dereferences a null pointer.
+/// The `expr` in `addr_of!(expr)` is evaluated as a place expression, but never loads
+/// from the place or requires the place to be dereferenceable. This means that
+/// `addr_of!(*ptr)` is defined behavior even if `ptr` is dangling or misaligned.
```
Comment (cuviper) [inline library/core/src/ptr/mod.rs:1994]: Now it seems like we've forgotten `null`, but this is allowed too, right?
```suggestion
/// `addr_of!(*ptr)` is defined behavior even if `ptr` is null, dangling, or misaligned.
```

File: library/core/src/ptr/mod.rs [hunk 2]
```diff
@@ -2029,9 +2038,18 @@ pub macro addr_of($place:expr) {
 /// as all other references. This macro can create a raw pointer *without* creating
 /// a reference first.
 ///
-/// Note, however, that the `expr` in `addr_of_mut!(expr)` is still subject to all
-/// the usual rules. In particular, `addr_of_mut!(*ptr::null_mut())` is Undefined
-/// Behavior because it dereferences a null pointer.
+/// The `expr` in `addr_of_mut!(expr)` is evaluated as a place expression, but never loads
+/// from the place or requires the place to be dereferenceable. This means that
+/// `addr_of_mut!(*ptr)` is defined behavior even if `ptr` is dangling or misaligned.
```
Comment (cuviper) [inline library/core/src/ptr/mod.rs:2043]: Ditto
```suggestion
/// `addr_of_mut!(*ptr)` is defined behavior even if `ptr` is null, dangling, or misaligned.
```

File: library/core/src/ptr/mod.rs (hunk 1 above)
Comment (RalfJung) [reply library/core/src/ptr/mod.rs:1994]: Sure, I was thinking "null" is just a special case of "dangling", but we can also list it explicitly.

**Pull request merged**